### ISSUE TITLE
Invalid SWIG code is generated for a template that's defined and used in the same file

### ISSUE THREAD

**Issue by dbankieris** [nasa/trick#1704]
Another problem resulting from #1679.

```c++
namespace a { 

template <class T>
class Bar {}; 

class Foo {
  public:
    Bar<int> bar;
};

}
```

Compiling yields `Foo_py.i:21: Error: Template 'Bar' undefined.`

This is due to `%template` being placed before the definition of `Bar` in `Foo_py.i`

```c++
#ifndef TRICK_SWIG_TEMPLATE_a__Foo_bar
#define TRICK_SWIG_TEMPLATE_a__Foo_bar
namespace a { 
%template(a__Foo_bar) Bar<int>;
}
#endif

namespace a{

template <class T>
class Bar {};
```

**Comment by dbankieris** [comment]
Also fails if the definition of `Bar` is outside of the `a` namespace, although for a different reason.
```c++
template <class T>
class Bar {}; 

namespace a { 

  class Foo {
    public:
      Bar<int> bar;
  };  

}
```

Compiling yields `Foo_py.i:26: Error: 'Bar' resolves to '::Bar' and was incorrectly instantiated in scope 'a' instead of within scope ''.`

This is due to `%template` being placed in the wrong namespace.

```c++
template <class T>
class Bar {}; 

#ifndef TRICK_SWIG_TEMPLATE_a__Foo_bar
#define TRICK_SWIG_TEMPLATE_a__Foo_bar
namespace a { 
%template(a__Foo_bar) Bar<int>;
}
#endif
```

**Comment by hchen99** [comment]
Thanks for reporting. We'll look into this.

**Comment by sharmeye** [comment]
We think this is addressed by #1741.  Closing this down.
